File: src/main/java/io/spring/core/article/Article.java
package io.spring.core.article;

import java.time.Instant;
import java.util.HashSet;
import java.util.List;
import java.util.UUID;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
@EqualsAndHashCode(of = {"id"})
public class Article {
  private String userId;
  private String id;
  private String slug;
  private String title;
  private String description;
  private String body;
  private List<Tag> tags;
  private Instant createdAt;
  private Instant updatedAt;

  public Article(
      String title, String description, String body, List<String> tagList, String userId) {
    this(title, description, body, tagList, userId, Instant.now());
  }

  public Article(
      String title,
      String description,
      String body,
      List<String> tagList,
      String userId,
      Instant createdAt) {
    this.id = UUID.randomUUID().toString();
    this.slug = toSlug(title);
    this.title = title;
    this.description = description;
    this.body = body;
    this.tags = new HashSet<>(tagList).stream().map(Tag::new).toList();
    this.userId = userId;
    this.createdAt = createdAt;
    this.updatedAt = createdAt;
  }

  public void update(String title, String description, String body) {
    if (title != null && !title.isEmpty()) {
      this.title = title;
      this.slug = toSlug(title);
      this.updatedAt = Instant.now();
    }
    if (description != null && !description.isEmpty()) {
      this.description = description;
      this.updatedAt = Instant.now();
    }
    if (body != null && !body.isEmpty()) {
      this.body = body;
      this.updatedAt = Instant.now();
    }
  }

  public static String toSlug(String title) {
    return title.toLowerCase().replaceAll("[\\&|[\\uFE30-\\uFFA0]|\\’|\\”|\\s\\?\\,\\.]+", "-");
  }
}

File: python/conduit/core/article.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional, Protocol

from conduit.core.util import is_empty, new_id, now_utc, to_slug


class Tag:
    def __init__(self, name: str = "", id: Optional[str] = None) -> None:
        self.id = id if id is not None else new_id()
        self.name = name

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Tag) and other.name == self.name

    def __hash__(self) -> int:
        return hash(self.name)


class Article:
    def __init__(
        self,
        title: str = "",
        description: str = "",
        body: str = "",
        tag_list: Optional[List[str]] = None,
        user_id: str = "",
        created_at: Optional[datetime] = None,
        id: Optional[str] = None,
    ) -> None:
        self.id = id if id is not None else new_id()
        self.title = title
        self.slug = to_slug(title)
        self.description = description
        self.body = body
        # de-duplicate tags while preserving Tag semantics
        unique_names: List[str] = []
        for name in tag_list or []:
            if name not in unique_names:
                unique_names.append(name)
        self.tags: List[Tag] = [Tag(name) for name in unique_names]
        self.user_id = user_id
        created = created_at if created_at is not None else now_utc()
        self.created_at = created
        self.updated_at = created

    def update(self, title: str, description: str, body: str) -> None:
        if not is_empty(title):
            self.title = title
            self.slug = to_slug(title)
            self.updated_at = now_utc()
        if not is_empty(description):
            self.description = description
            self.updated_at = now_utc()
        if not is_empty(body):
            self.body = body
            self.updated_at = now_utc()

    @staticmethod
    def to_slug(title: str) -> str:
        return to_slug(title)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Article) and other.id == self.id

    def __hash__(self) -> int:
        return hash(self.id)


class ArticleRepository(Protocol):
    def save(self, article: Article) -> None: ...

    def find_by_id(self, id: str) -> Optional[Article]: ...

    def find_by_slug(self, slug: str) -> Optional[Article]: ...

    def remove(self, article: Article) -> None: ...
